pkg/tecdsa/dkg/states.go

```diff
@@ -416,167 +301,93 @@ func (trts *tssRoundThreeState) MemberIndex() group.MemberIndex {
 type finalizationState struct {
+	*state.BaseAsyncState
+
 	channel net.BroadcastChannel
 	member  *finalizingMember
-
-	outcomeChan chan error
-
-	previousPhaseMessages []*tssRoundThreeMessage
-
-	phaseMessages []*tssFinalizationMessage
-}
-
-func (fs *finalizationState) DelayBlocks() uint64 {
-	return finalizationStateDelayBlocks
-}
-
-func (fs *finalizationState) ActiveBlocks() uint64 {
-	return finalizationStateActiveBlocks
 }
 
 func (fs *finalizationState) Initiate(ctx context.Context) error {
-	fs.member.markInactiveMembers(fs.previousPhaseMessages)
-
-	if len(fs.member.group.InactiveMemberIDs()) > 0 {
-		return newInactiveMembersError(fs.member.group.InactiveMemberIDs())
+	message, err := fs.member.tssFinalize(
+		ctx,
+		receivedMessages[*tssRoundThreeMessage](fs.BaseAsyncState),
+	)
+	if err != nil {
+		return err
 	}
 
-	// TSS computations can be time-consuming and can exceed the current
-	// state's time window. The ctx parameter is scoped to the lifetime of
-	// the current state so, it can be used as a timeout signal. However,
-	// that ctx is cancelled upon state's end only after Initiate returns.
-	// In order to make that working, Initiate must trigger the computations
-	// in a separate goroutine and return before the end of the state.
-	go func() {
-		message, err := fs.member.tssFinalize(ctx, fs.previousPhaseMessages)
-		if err != nil {
-			fs.outcomeChan <- err
-			return
-		}
-
-		if err := fs.channel.Send(ctx, message); err != nil {
-			fs.outcomeChan <- err
-			return
-		}
-
-		close(fs.outcomeChan)
-	}()
+	if err := fs.channel.Send(
+		ctx,
+		message,
+		net.BackoffRetransmissionStrategy,
+	); err != nil {
+		return err
+	}
 
 	return nil
 }
 
-func (fs *finalizationState) Receive(msg net.Message) error {
-	switch phaseMessage := msg.Payload().(type) {
-	case *tssFinalizationMessage:
+func (fs *finalizationState) Receive(netMessage net.Message) error {
+	if protocolMessage, ok := netMessage.Payload().(message); ok {
 		if fs.member.shouldAcceptMessage(
-			phaseMessage.SenderID(),
-			msg.SenderPublicKey(),
-		) && fs.member.sessionID == phaseMessage.sessionID {
-			fs.phaseMessages = append(fs.phaseMessages, phaseMessage)
+			protocolMessage.SenderID(),
+			netMessage.SenderPublicKey(),
+		) && fs.member.sessionID == protocolMessage.SessionID() {
+			fs.ReceiveToHistory(netMessage)
 		}
 	}
 
 	return nil
 }
 
-func (fs *finalizationState) Next() (state.SyncState, error) {
-	err := <-fs.outcomeChan
-	if err != nil {
-		return nil, err
-	}
-
-	return &confirmationState{
-		channel:               fs.channel,
-		member:                fs.member.initializeConfirmation(),
-		previousPhaseMessages: fs.phaseMessages,
-	}, nil
-}
-
-func (fs *finalizationState) MemberIndex() group.MemberIndex {
-	return fs.member.id
-}
-
-// confirmationState is the state during which members confirm a successful
-// course of the distributed key generation process.
-type confirmationState struct {
```

Comment:
Checking my understanding of why we can remove `confirmationState`.

From #3340:

> The problematic case is when members fail at the last phase of DKG (finalization round) which is not followed by any communication round allowing for detection of such failures. In that scenario, members who failed the last round do not produce the key material but, at the same time, are not excluded by other participants as misbehaved, so they can be selected for signing normally. As a result, signature attempts including those members fail and the wallet needs more attempts in total to exclude those members using the random algorithm and come up with a successful signature.

Some members could fail to execute `fs.member.tssFinalize` in time and the group did not know about it. In #3347 we introduced `tssFinalizationMessage` and one more phase after `finalizationState` to confirm enough group members sent `tssFinalizationMessage`.

That phase after `finalizationState` is no longer needed because `CanTransition()` will block until `fs.member.group.OperatingMemberIDs()` `tssFinalizationMessage`s are received. This can be translated to: everyone who participated in DKG completed computations in `fs.member.tssFinalize`.

Does it make sense?

Reply:
Yep, exactly!